File: solutions/projectstaffing/jgraph/core/src/main/scala/com/microsoft/graphdataconnect/skillsfinder/models/business/employee/Employee.scala
package com.microsoft.graphdataconnect.skillsfinder.models.business.employee

import com.microsoft.graphdataconnect.skillsfinder.utils.SeqUtils.ListImplicits
import com.microsoft.graphdataconnect.skillsfinder.utils.StringUtils.StringImplicits

// TODO remove the nulls. Instead use proper values in all constructor calls
case class Employee(id: String,
                    name: String,
                    mail: String,
                    role: String,
                    inferredRoles: List[String],
                    location: String,
                    about: String = null,
                    relevantSkills: List[String],
                    inferredSkills: List[String] = null,
                    declaredSkills: List[String],
                    domainToSkillMap: Map[String, List[String]],
                    highlightedTerms: List[String] = null,
                    topics: List[String] = null,
                    currentEngagement: String = null,
                    department: String = null,
                    reportsTo: String = null,
                    managerEmail: String = null,
                    availableSince: String = null,
                    linkedInProfile: String,
                    includedInCurrentTeam: Option[Boolean] = None,
                    profilePicture: String)

object Employee {
  def dropEmptyElementsFromLists(employee: Employee): Employee = {
    employee.copy(
      inferredRoles = employee.inferredRoles.nullToNil.filterNot(_.isNullOrBlank),
      declaredSkills = employee.declaredSkills.nullToNil.filterNot(_.isNullOrBlank),
      relevantSkills = employee.relevantSkills.nullToNil.filterNot(_.isNullOrBlank),
      inferredSkills = employee.inferredSkills.nullToNil.filterNot(_.isNullOrBlank),
      highlightedTerms = employee.highlightedTerms.nullToNil.filterNot(_.isNullOrBlank)
    )
  }
}

File: solutions/projectstaffing/jgraph/ui/src/types/Employee.ts
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project root for full license information.
 */

import store from '@/store';

export function isSkillSearched(terms: string[], skill: string) {
  return (
    terms.find((sk: string) => {
      return skill.toLowerCase() === sk.toLowerCase();
    }) !== undefined
  );
}

export class Employee {
  public id?: string;
  public name?: string = '';
  public department?: string;
  public mail?: string;
  public role?: string;
  public location?: string;
  public timeZone?: string;
  public timeNow?: string;
  public about?: string;
  public topics?: string[] = [];
  public inferredRoles?: string[] = [];
  public highlightedTerms?: string[] = [];
  public inferredSkills?: {
    highlighted: boolean;
    value: string;
  }[];
  public relevantSkills?: {
    highlighted: boolean;
    value: string;
  }[];
  public relevantSkillsHTML: any;
  public declaredSkills?: {
    highlighted: boolean;
    value: string;
  }[];
  public domainToSkillMap: any;
  public domainToSkillMapHTML: any;
  public currentEngagement?: string;
  public reportsTo?: string;
  public availableSince?: string;
  public profilePicture?: string;
  public showMore?: boolean = true;
  public isSkillEllipsis?: boolean = false;
  public includedInCurrentTeam?: boolean = false;
  public managerEmail?: string = '';
  public linkedInProfile?: string = '';

  constructor(init?: Partial<Employee>) {
    Object.assign(this, init);
    if (init?.relevantSkills) {
      this.relevantSkills = [...new Set(init.relevantSkills)].map(
        (skill: any) => {
          return {
            highlighted: isSkillSearched(this.highlightedTerms!, skill),
            value: skill
          };
        }
      );
      this.relevantSkillsHTML =
        `<span class="mr-1 text-black text-r"
      >Profile Skills:</span
    >` +
        [...new Set(init.relevantSkills)]
          .map((skill: any, index: number) => {
            const isSkillHighlighted = isSkillSearched(
              this.highlightedTerms!,
              skill
            );
            if (isSkillHighlighted) {
              return ` <span
      class="skill highlighted-skill"
      >${skill}${
                index + 1 !== init.relevantSkills!.length
                  ? '<span class="text-secondary custom-comma">,</span>'
                  : ''
              }</span
    >`;
            } else {
              return `${skill}${
                index + 1 !== init.relevantSkills!.length ? ',' : ''
              } `;
            }
          })
          .join('');
    }
    if (init?.declaredSkills)
      this.declaredSkills = [...new Set(init.declaredSkills)].map(
        (skill: any) => {
          return {
            highlighted: isSkillSearched(this.highlightedTerms!, skill),
            value: skill
          };
        }
      );
    if (init?.inferredSkills)
      this.inferredSkills = [...new Set(init.inferredSkills)].map(
        (skill: any) => {
          return {
            highlighted: isSkillSearched(this.highlightedTerms!, skill),
            value: skill
          };
        }
      );
    if (init?.domainToSkillMap) {
      let empDomainToSkillMapHTML: any = {};
      Object.keys(init.domainToSkillMap).map(domain => {
        if (init.domainToSkillMap[domain].length !== 0) {
          const domainArr = [...new Set(init.domainToSkillMap[domain])].filter(
            x => x
          );
          if (domainArr.length !== 0)
            empDomainToSkillMapHTML[domain] =
              `  <span class="mr-1 text-black text-r"
            >${(store.state as any).team.taxonomies[domain]} Skills:</span
          >` +
              domainArr
                .map((skill: any, index: number) => {
                  const isSkillHighlighted = isSkillSearched(
                    this.highlightedTerms!,
                    skill
                  );
                  if (isSkillHighlighted) {
                    return ` <span
              class="skill highlighted-skill"
              >${skill}${
                      index + 1 !== domainArr.length
                        ? '<span class="text-secondary custom-comma">,</span>'
                        : ''
                    }</span
            >`;
                  } else {
                    return `${skill}${
                      index + 1 !== domainArr.length ? ',' : ''
                    } `;
                  }
                })
                .join('');
        }
      });
      this.domainToSkillMapHTML = empDomainToSkillMapHTML;
    }
  }
}

export class TeamMember {
  public userId?: string;
  public employeeId?: string;
  public teamMemberId?: string;
  public name?: string;
  public email?: string;
  public skills?: string[];
  public availableSince?: string;
  public employeeRole?: string;
  public employeeProfilePicture?: string;

  constructor(init?: Partial<TeamMember>) {
    Object.assign(this, init);
  }
}
